File: frontend/src/types/node.ts
export interface Node {
  id: number
  hostname: string
  ip_address: string | null
  status: 'ONLINE' | 'OFFLINE'
  last_heartbeat: string | null
  registered_at: string
  os: 'WINDOWS' | 'LINUX' | 'MACOS' | 'UNKNOWN'
  os_version: string | null
  cpu_cores: number | null
  cpu_usage: number | null
  memory_gb: number | null
  memory_usage: number | null
  disk_gb: number | null
  disk_usage: number | null
  version: string | null
  python_version: string | null
  tags: string | null
  capabilities: string | null
  max_concurrency: number
  current_concurrency: number
  public_ip: string | null
  agent_port: number | null
  physical_host_id: string | null
  physical_host_name: string | null
  container_id: string | null
  instance_name: string | null
}

export interface NodeCreate {
  hostname: string
  ip_address?: string
  os?: 'WINDOWS' | 'LINUX' | 'MACOS' | 'UNKNOWN'
  physical_host_id?: string
  physical_host_name?: string
  container_id?: string
  instance_name?: string
}

export interface NodeUpdate {
  ip_address?: string
  os?: 'WINDOWS' | 'LINUX' | 'MACOS' | 'UNKNOWN'
  os_version?: string
  cpu_cores?: number
  cpu_usage?: number
  memory_gb?: number
  memory_usage?: number
  disk_gb?: number
  disk_usage?: number
  version?: string
  python_version?: string
  tags?: string
  capabilities?: string
  max_concurrency?: number
  public_ip?: string
  agent_port?: number
  physical_host_id?: string
  physical_host_name?: string
  container_id?: string
  instance_name?: string
}

File: backend/app/models/node.py
# /backend/app/models/node.py

from typing import Optional, List
from enum import Enum as PyEnum
from sqlalchemy import Integer, String, DateTime, func, Enum as SqlEnum, Float, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.db.base_class import Base


class NodeStatus(str, PyEnum):
    ONLINE = "ONLINE"
    OFFLINE = "OFFLINE"


class NodeOS(str, PyEnum):
    """
    节点操作系统类型
    """
    WINDOWS = "WINDOWS"
    LINUX = "LINUX"
    MACOS = "MACOS"
    UNKNOWN = "UNKNOWN"


class Node(Base):
    __tablename__ = "cp_nodes"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    hostname: Mapped[str] = mapped_column(String(100), unique=True, index=True, nullable=False)
    ip_address: Mapped[Optional[str]] = mapped_column(String(50))
    status: Mapped[NodeStatus] = mapped_column(
        SqlEnum(NodeStatus, name="node_status_enum"),
        default=NodeStatus.OFFLINE,
        nullable=False
    )
    last_heartbeat: Mapped[Optional[DateTime]] = mapped_column(DateTime)
    registered_at: Mapped[DateTime] = mapped_column(DateTime, server_default=func.now())

    # ✅ 操作系统信息（用于区分 Windows/Linux）
    os: Mapped[NodeOS] = mapped_column(
        SqlEnum(NodeOS, name="node_os_enum"),
        default=NodeOS.UNKNOWN,
        nullable=False
    )
    os_version: Mapped[Optional[str]] = mapped_column(String(50), comment="操作系统版本，如 Windows 11, Ubuntu 20.04")

    # ✅ 硬件资源
    cpu_cores: Mapped[Optional[int]] = mapped_column(Integer, comment="CPU 核心数")
    cpu_usage: Mapped[Optional[float]] = mapped_column(Float, comment="CPU 使用率 (%)")
    memory_gb: Mapped[Optional[float]] = mapped_column(Float, comment="内存大小（GB）")
    memory_usage: Mapped[Optional[float]] = mapped_column(Float, comment="内存使用量（GB）")
    disk_gb: Mapped[Optional[float]] = mapped_column(Float, comment="磁盘大小（GB）")
    disk_usage: Mapped[Optional[float]] = mapped_column(Float, comment="磁盘使用量（GB）")

    # ✅ 软件与版本
    version: Mapped[Optional[str]] = mapped_column(String(20), comment="Worker 版本")
    python_version: Mapped[Optional[str]] = mapped_column(String(20), comment="Python 版本")
    # 如果支持 Node.js/PHP 等，也可加

    # ✅ 标签与能力
    tags: Mapped[Optional[str]] = mapped_column(String(100), comment="节点标签，如 gpu,proxy,chrome")
    capabilities: Mapped[Optional[str]] = mapped_column(
        Text,
        comment="节点能力（JSON 格式），如 {'gpu': true, 'browser': 'chrome'}"
    )  # 后续可改为 JSON 类型

    # ✅ 并发控制
    max_concurrency: Mapped[int] = mapped_column(Integer, default=4, comment="最大并发任务数")
    current_concurrency: Mapped[int] = mapped_column(Integer, default=0, comment="当前运行任务数")

    # ✅ 网络信息（可选）
    public_ip: Mapped[Optional[str]] = mapped_column(String(50), comment="公网 IP")
    agent_port: Mapped[Optional[int]] = mapped_column(Integer, comment="Worker 通信端口")
    
    # ✅ 物理主机信息（用于支持同一物理机上的多个逻辑节点）
    physical_host_id: Mapped[Optional[str]] = mapped_column(String(100), comment="物理主机标识（如MAC地址）")
    physical_host_name: Mapped[Optional[str]] = mapped_column(String(100), comment="物理主机名")
    container_id: Mapped[Optional[str]] = mapped_column(String(100), comment="容器ID（如果是Docker容器）")
    instance_name: Mapped[Optional[str]] = mapped_column(String(100), comment="实例名称（用户自定义）")